Platform/ToggleSipEntry/ToggleSipEntry.c

```diff
@@ -0,0 +1,172 @@
+    .Tool          = FALSE,
+    .TextMode      = FALSE,
+    .RealPath      = FALSE,
+    .SystemAction  = InternalSystemActionSetSip,
+    .ActionConfig  = NULL, ///< in this driver we use static variables, instead
+    .AudioBasePath = NULL, ///< overridden in set up
+    .AudioBaseType = OC_VOICE_OVER_AUDIO_BASE_TYPE_OPEN_CORE
+  }
+};
+
+STATIC
+EFI_STATUS
+EFIAPI
+ToggleSipGetBootEntries (
+  IN           OC_PICKER_CONTEXT  *PickerContext,
+  IN     CONST EFI_HANDLE         Device,
+  OUT       OC_PICKER_ENTRY       **Entries,
+  OUT       UINTN                 *NumEntries
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      CsrActiveConfig;
+  BOOLEAN     IsEnabled;
+
+  //
+  // Custom entries only.
+  //
+  if (Device != NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  //
+  // Configure current action.
+  //
+  Status = OcGetSip (&CsrActiveConfig, &mAttributes);
+  if (!EFI_ERROR (Status) || (Status == EFI_NOT_FOUND)) {
+    IsEnabled = OcIsSipEnabled (Status, CsrActiveConfig);
+  } else {
+    DEBUG ((DEBUG_WARN, "BEP: ToggleSip failed to read csr-active-config, aborting! - %r\n", Status));
+    return Status;
+  }
+
+  if (IsEnabled) {
+    mToggleSipBootEntries[0].Name          = OC_MENU_SIP_IS_ENABLED;
+    mToggleSipBootEntries[0].Flavour       = OC_FLAVOUR_TOGGLE_SIP_ENABLED;
+    mToggleSipBootEntries[0].AudioBasePath = OC_VOICE_OVER_AUDIO_FILE_SIP_IS_ENABLED;
+    mCsrNextConfig                         = mCsrUserConfig;
+  } else {
+    mToggleSipBootEntries[0].Name          = OC_MENU_SIP_IS_DISABLED;
+    mToggleSipBootEntries[0].Flavour       = OC_FLAVOUR_TOGGLE_SIP_DISABLED;
+    mToggleSipBootEntries[0].AudioBasePath = OC_VOICE_OVER_AUDIO_FILE_SIP_IS_DISABLED;
+    mCsrNextConfig                         = 0;
+  }
+
+  DEBUG ((DEBUG_INFO, "BEP: Adding Toggle SIP entry, currently enabled %d, will change 0x%X->0x%X\n", IsEnabled, CsrActiveConfig, mCsrNextConfig));
+
+  *Entries    = mToggleSipBootEntries;
+  *NumEntries = ARRAY_SIZE (mToggleSipBootEntries);
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+OC_BOOT_ENTRY_PROTOCOL
+  mToggleSipBootEntryProtocol = {
+  OC_BOOT_ENTRY_PROTOCOL_REVISION,
+  ToggleSipGetBootEntries,
+  NULL
+};
+
+EFI_STATUS
+EFIAPI
+UefiMain (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS                 Status;
+  EFI_LOADED_IMAGE_PROTOCOL  *LoadedImage;
+  UINTN                      Data;
+
+  Status = gBS->HandleProtocol (
+                  ImageHandle,
+                  &gEfiLoadedImageProtocolGuid,
+                  (VOID **)&LoadedImage
+                  );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = EFI_NOT_FOUND;
+
+  //
+  // Is CHAR_NULL terminated string present?
+  //
+  if (  (LoadedImage->LoadOptions != NULL)
+     && ((LoadedImage->LoadOptionsSize % 2) == 0)
+     && ((LoadedImage->LoadOptionsSize / 2) > 0)
+     && (((CHAR16 *)LoadedImage->LoadOptions)[(LoadedImage->LoadOptionsSize / 2) - 1] == CHAR_NULL)
```

Comment:
Hard for me to check right now, but isn't such logic used elsewhere too? Could try to remember to move it to a library